File: Sources/ClaudePet/ClaudeMonitor.swift
import Foundation

enum ClaudeStatus: Equatable {
    case notRunning
    case idle
    case working
    case waitingForPermission
}

struct SessionInfo: Equatable {
    let sessionId: String
    let status: ClaudeStatus
    let cwd: String
    let tool: String
    let timestamp: Int

    static func == (lhs: SessionInfo, rhs: SessionInfo) -> Bool {
        return lhs.sessionId == rhs.sessionId && lhs.status == rhs.status
    }
}

class ClaudeMonitor {
    private let staleTimeout: Int = 7200 // 2시간 이상 업데이트 없으면 정리

    func checkSessions() -> [SessionInfo] {
        var sessions: [SessionInfo] = []
        let fileManager = FileManager.default

        guard let files = try? fileManager.contentsOfDirectory(atPath: "/tmp") else {
            return sessions
        }

        let now = Int(Date().timeIntervalSince1970)

        for file in files {
            guard file.hasPrefix("claudepet-") && file.hasSuffix(".json") else { continue }

            let path = "/tmp/\(file)"
            guard let data = fileManager.contents(atPath: path),
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let event = json["status"] as? String,
                  let sessionId = json["session_id"] as? String,
                  let ts = json["ts"] as? Int else { continue }

            // 오래된 상태 파일 정리
            if now - ts > staleTimeout {
                try? fileManager.removeItem(atPath: path)
                continue
            }

            let cwd = json["cwd"] as? String ?? ""
            let tool = json["tool"] as? String ?? "none"

            let status: ClaudeStatus
            switch event {
            case "UserPromptSubmit", "PreToolUse":
                status = .working
            case "PostToolUse", "SubagentStop":
                // 도구 하나 끝났지만 아직 작업 중일 수 있음
                status = .working
            case "Stop":
                // Claude가 응답 완료 → 진짜 작업 끝
                status = .idle
            case "PermissionRequest":
                status = .waitingForPermission
            case "SessionStart":
                status = .idle
            default:
                status = .idle
            }

            sessions.append(SessionInfo(
                sessionId: sessionId,
                status: status,
                cwd: cwd,
                tool: tool,
                timestamp: ts
            ))
        }

        return sessions
    }
}

File: windows/src/main/claude-monitor.ts
/**
 * ClaudeMonitor.swift 포팅
 *
 * %TEMP%\claudepet-*.json 파일을 스캔해 세션 상태를 추적한다.
 * Hook 스크립트가 이 파일들을 작성한다.
 */

import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';

export type ClaudeStatus = 'notRunning' | 'idle' | 'working' | 'waitingForPermission';

export interface SessionInfo {
  sessionId: string;
  status: ClaudeStatus;
  cwd: string;
  tool: string;
  timestamp: number;
}

// Stale timeout — 비정상 종료된 세션 정리용.
// 정상 exit 시엔 SessionEnd hook 이 즉시 temp 파일을 지우므로 이 값과 무관.
// 강제 종료 / 크래시 / PC 종료 등 비정상 케이스만 이 timeout 으로 정리.
// 맥 버전(ClaudeMonitor.swift)과 동일하게 2시간으로 맞춤.
const STALE_TIMEOUT_SEC = 7200; // 2시간

export class ClaudeMonitor {
  private tmpDir: string;

  constructor() {
    this.tmpDir = os.tmpdir();
  }

  /** 현재 살아있는 세션 목록 반환 */
  checkSessions(): SessionInfo[] {
    const sessions: SessionInfo[] = [];

    let files: string[];
    try {
      files = fs.readdirSync(this.tmpDir);
    } catch {
      return sessions;
    }

    const now = Math.floor(Date.now() / 1000);

    for (const file of files) {
      if (!file.startsWith('claudepet-') || !file.endsWith('.json')) continue;

      const filePath = path.join(this.tmpDir, file);
      let data: any;
      try {
        let content = fs.readFileSync(filePath, 'utf8');
        // PowerShell 5.x 의 Out-File -Encoding utf8 은 BOM 을 붙임 → 제거
        if (content.charCodeAt(0) === 0xFEFF) {
          content = content.slice(1);
        }
        data = JSON.parse(content);
      } catch {
        continue;
      }

      const event: string = data.status || '';
      const sessionId: string = data.session_id || '';
      const ts: number = data.ts || 0;
      if (!sessionId || !ts) continue;

      // session_id 추출 실패한 케이스는 무시 + 정리
      if (sessionId === 'unknown') {
        try {
          fs.unlinkSync(filePath);
        } catch {}
        continue;
      }

      // 오래된 파일 정리 (Claude Code 가 SessionEnd 안 보내고 죽은 경우)
      if (now - ts > STALE_TIMEOUT_SEC) {
        try {
          fs.unlinkSync(filePath);
        } catch {}
        continue;
      }

      const cwd: string = data.cwd || '';
      const tool: string = data.tool || 'none';

      let status: ClaudeStatus;
      switch (event) {
        case 'UserPromptSubmit':
        case 'PreToolUse':
        case 'PostToolUse':
        case 'SubagentStop':
          status = 'working';
          break;
        case 'Stop':
          status = 'idle';
          break;
        case 'PermissionRequest':
          status = 'waitingForPermission';
          break;
        case 'SessionStart':
          status = 'idle';
          break;
        default:
          status = 'idle';
          break;
      }

      sessions.push({ sessionId, status, cwd, tool, timestamp: ts });
    }

    return sessions;
  }

  /**
   * Claude Desktop 프로세스가 실행 중인지 확인.
   *
   * 주의: Windows 에선 Claude Code CLI 도 `claude.exe` 라는 이름으로 떠있다.
   * 따라서 단순 프로세스 이름 매칭만으론 구분이 안 된다.
   *
   * 해결: PowerShell 로 ExecutablePath 까지 가져와서, 경로에
   * "AnthropicClaude" (Desktop 설치 경로) 가 포함된 프로세스만 카운트.
   *
   *   - Claude Code CLI:  C:\Users\...\npm\node_modules\@anthropic-ai\claude-code\...
   *   - Claude Desktop:   C:\Users\...\AppData\Local\AnthropicClaude\Claude.exe (또는 Program Files)
   */
  async isDesktopRunning(): Promise<boolean> {
    return new Promise((resolve) => {
      const { execFile } = require('child_process');
      const ps = `Get-CimInstance Win32_Process -Filter "name='claude.exe'" | ` +
                 `Where-Object { $_.ExecutablePath -like '*AnthropicClaude*' -or $_.ExecutablePath -like '*Anthropic Claude*' } | ` +
                 `Select-Object -First 1 -ExpandProperty ProcessId`;
      execFile(
        'powershell',
        ['-NoProfile', '-ExecutionPolicy', 'Bypass', '-Command', ps],
        { windowsHide: true },
        (err: any, stdout: string) => {
          if (err) {
            resolve(false);
            return;
          }
          const trimmed = (stdout || '').trim();
          resolve(/^\d+/.test(trimmed));
        }
      );
    });
  }
}
